### PR TITLE
Semaphore implementation without IPC

### PR DESCRIPTION
**Pull request by lebrush**
This is a working semaphore implementation using mutex and thread_flags. It seems to work fine but firstly this is a request for comments.

(fix for #6151)

Depends on: #6158

### CODE REVIEW

File: sys/include/xtimer.h [hunk 1]
```diff
@@ -279,6 +280,19 @@ int xtimer_msg_receive_timeout(msg_t *msg, uint32_t us);
...
+ *
+ * @note this requires core_thread_flags to be enabled
+ *
+ * @param[in]    mutex  mutex to lock
+ * @param[in]    us     timeout in microseconds relative
+ *
+ * @return       0, when returned after mutex was locked
+ * @return       -1, when the timeout occcured
+ */
+int xtimer_mutex_lock_timeout(mutex_t *mutex, uint64_t us);
```
Comment (miri64) [inline sys/include/xtimer.h]: I would prefer this change to xtimer to be introduced in a separate PR.

File: sys/include/sema.h [hunk 2]
```diff
@@ -33,21 +35,37 @@ extern "C" {
...
+    sema_state_t state;             /**< state of the semaphore */
+    mutex_t mutex;                  /**< mutex of the semaphore */
 } sema_t;
 
 /**
+ * @brief   Get value of a semaphore.
+ *
+ * @return  Current value of the semaphore.
+ */
+#define SEMA_VALUE(sema)        ATOMIC_VALUE(sema->value)
```
Comment (miri64) [inline sys/include/sema.h]: Can this be a inline function instead.

File: sys/include/sema.h [hunk 3]
```diff
@@ -33,21 +35,37 @@ extern "C" {
...
 
 /**
  * @brief A Semaphore.
  */
 typedef struct {
-    volatile unsigned int value;    /**< value of the semaphore */
-    priority_queue_t queue;         /**< list of threads waiting for the semaphore */
+    atomic_int_t value;             /**< value of the semaphore */
+    sema_state_t state;             /**< state of the semaphore */
+    mutex_t mutex;                  /**< mutex of the semaphore */
```
Comment (miri64) [inline sys/include/sema.h]: Do you really need thread flags *and* mutex?

**Comment by lebrush** [comment]
I found two race conditions after creating the PR. I'll tackle them tomorrow.

**Comment by Kijewski** [comment]
I am inclined to remove the "enhancement" label because you removed the `int sema_wait_timed_msg(sema_t *sema, uint64_t timeout, msg_t *msg)` functionality with no replacement.

**Comment by lebrush** [comment]
@Kijewski good point. The same situation could be claimed however for the mutex: what if you receive a message while in a mutex? both mutex and semaphore are concurrency control mechanisms. When you wait in your code for something (i.e. mutex) you wait for it, the messages are going to be queued.

In this case I modified the sema implementation so it doesn't use messages anymore, hence the `sema_wait_timed_msg` (not used anywhere in the code btw, except by lwip) is not applying anymore.

File: sys/include/sema.h (hunk 3 above)
Comment (lebrush) [inline sys/include/sema.h]: Absolutely right, I removed the flags dependency :+1:

File: sys/include/sema.h (hunk 2 above)
Comment (lebrush) [inline sys/include/sema.h]: done :+1:

File: sys/include/xtimer.h (hunk 1 above)
Comment (lebrush) [inline sys/include/xtimer.h]: Might make sense, it's in a separate commit in any case :-)

**Comment by lebrush** [comment]
@miri64 I fixed the race conditions and: 
 - removed dependency to atomic
 - removed dependency to thread_flags

File: sys/include/sema.h (hunk 3 above)
Comment (miri64) [inline sys/include/sema.h]: mh.... actually I meant it the other way around. `sema` used to be based on flags (or rather `posix_semaphore` was before we took the base implementation out of the POSIX wrapper and made it message-based) and it didn't work out as we thought it might (see also size of the current `sema_t` struct vs how it was before).

Comment (lebrush) [inline sys/include/sema.h]: Before it was `unsigned int` + `priority_queue_node_t`, assuming a 32-bit processor this is `4 + 12 = 16` byte. Now it is `unsigned int` + `enum` + `mutex_t` which is `4+4+4 = 12` bytes. This is 4 byte less. One could remove the `enum` and use `signed int` for the value (negative meaning destroy) and could save 4 bytes more (the posix implementation uses a signed int).

I liked the point to remove the flags dependency because, adding the flags adds 2 bytes to every thread stack even if the semaphore is only used by some of the threads.

File: sys/include/sema.h [hunk 4]
```diff
@@ -69,57 +89,25 @@ int sema_create(sema_t *sema, unsigned int value);
...
- * @param[in]  timeout  Time in microseconds until the semaphore times out. 0 for no timeout.
- * @param[out] msg      Container for a spurious message during the timed wait (result == -EAGAIN).
- *
- * @return  0 on success
- * @return  -EINVAL, if semaphore is invalid.
- * @return  -ETIMEDOUT, if the semaphore times out.
- * @return  -ECANCELED, if the semaphore was destroyed.
- * @return  -EAGAIN, if the thread received a message while waiting for the lock.
- */
-int sema_wait_timed_msg(sema_t *sema, uint64_t timeout, msg_t *msg);
```
Comment (miri64) [inline sys/include/sema.h]: If you remove that: Doesn't lwIP needs fixing, too?

File: sys/include/sema.h [hunk 5]
```diff
@@ -129,22 +117,28 @@ int sema_wait_timed(sema_t *sema, uint64_t timeout);
  * @param[in]  sema A semaphore.
  *
  * @return  0 on success
- * @return  -EINVAL, if semaphore is invalid.
- * @return  -ECANCELED, if the semaphore was destroyed.
+ * @return  -1 on error (i.e. semaphore was destroyed)
```
Comment (miri64) [inline sys/include/sema.h]: Here and in other places: Changing these return values changes the behavior of the POSIX wrapper (it uses the result to set `errno`, as specified by the POSIX specifications). Either we drop this behavior (and don't care about POSIX compatibility and also fix the POSIX wrapper accordingly) or these values needs to stay the same.

File: sys/include/sema.h (hunk 3 above)
Comment (miri64) [inline sys/include/sema.h]: Okay, I somehow thought `priority_queue_node_t` was just 4 byte...

File: sys/include/sema.h (hunk 4 above)
Comment (lebrush) [inline sys/include/sema.h]: No, lwIP was calling `sema_wait_timed` which was calling `sema_wait_timed_msg`. However, your next point about the return values makes that lwIP requires fixing, I'll rollback the return values to the original ones.

File: sys/include/sema.h (hunk 5 above)
Comment (lebrush) [inline sys/include/sema.h]: I restored the "old" return codes :+1:

**Comment by miri64** [comment]
@Kijewski 

> I am inclined to remove the "enhancement" label because you removed the int sema_wait_timed_msg(sema_t *sema, uint64_t timeout, msg_t *msg) functionality with no replacement. What to do when I get a message while waiting for a semaphore? Who says that the semaphore is more important than a message?

This PR removes the dependency of IPC from sema, so you won't receive a message while waiting for a semaphore. Did I get this right @lebrush?

**Comment by lebrush** [comment]
Right

File: sys/include/xtimer.h (hunk 1 above)
Comment (miri64) [inline sys/include/xtimer.h]: Yes, but from a review perspective it makes even more sense. The know-how about semaphores and xtimer can be assumed to be split among the reviewers, so this PR would essentially need to ACKs: one from a person knowing about xtimer and one about semaphores ;-).

**Comment by lebrush** [comment]
@miri64 @Kijewski this has been now split into two PR this one tackling the Semaphore and #6158 for the mutex implementation with timeout. This one will not build without #6158

File: sys/include/sema.h [hunk 6]
```diff
@@ -33,21 +34,42 @@ extern "C" {
...
 } sema_t;
 
 /**
+ * @brief   Get value of a semaphore.
+ *
+ * @param[in]  sema Semaphore to get the value from
+ *
+ * @return  Current value of the semaphore.
+ */
+static inline unsigned int sema_value(sema_t *sema)
```
Comment (miri64) [inline sys/include/sema.h]: Why is this prefered over just taking `sema->value` (either remove function or rationalize in Doxygen).

File: sys/sema/sema.c [hunk 7]
```diff
@@ -11,169 +12,109 @@
  *
  * @file
  *
- * @author  Christian Mehlis <mehlis@inf.fu-berlin.de>
- * @author  Martine Lenders <mlenders@inf.fu-berlin.de>
- * @author  René Kijewski <kijewski@inf.fu-berlin.de>
+ * @author  Víctor Ariño <victor.arino@zii.aero>
```
Comment (miri64) [inline sys/sema/sema.c]: Keeping us in the Copyright but removing our authorship? ;-)

(though given the amount of code left I understand the latter step)

File: sys/include/sema.h (hunk 6 above)
Comment (lebrush) [inline sys/include/sema.h]: Actually it doesn't make a difference, so I removed it :+1:  Then there was no need to touch the `semaphore.h` file :)

File: sys/include/sema.h [hunk 8]
```diff
@@ -18,13 +18,14 @@
  * @author  Martine Lenders <mlenders@inf.fu-berlin.de>
  * @author  Christian Mehlis <mehlis@inf.fu-berlin.de>
  * @author  René Kijewski <kijewski@inf.fu-berlin.de>
+ * @author  Víctor Ariño <victor.arino@zii.aero>
  */
 #ifndef SEM_H_
```
Comment (miri64) [inline sys/include/sema.h]: Minor legacy error that might be fixable in this PR: `s/SEM_H_/SEMA_H_/g`

File: sys/include/sema.h [hunk 9]
```diff
@@ -33,18 +34,27 @@ extern "C" {
 /**
  * @brief   Creates semaphore statically.
  *
- * @param[in] value Initial value for the semaphore.
+ * The value for the statically created semaphore must be 0
  *
  * @return  Statically initialized semaphore.
  */
-#define SEMA_CREATE(value)      { (value), PRIORITY_QUEUE_INIT }
+#define SEMA_CREATE()           { (0), SEMA_OK, MUTEX_INIT() }
```
Comment (miri64) [inline sys/include/sema.h]: I see benefit in initializing a semaphore with a non-zero value e.g. to let more than one thread pass initially (furthermore, you also kept it for dynamic-equivalent `sema_create()`). So why change it? Or why *must* it be 0 (as per documentation).

File: sys/include/sema.h [hunk 10]
```diff
@@ -58,7 +68,6 @@ typedef struct {
  * @param[in] value Initial value for the semaphore.
  *
  * @return  0 on success.
- * @return  -EINVAL, if semaphore is invalid.
```
Comment (miri64) [inline sys/include/sema.h]: Here and below: please document the assertions you make in the implementation using the `@pre` doxygen command.

File: sys/sema/sema.c (hunk 7 above)
Comment (lebrush) [inline sys/sema/sema.c]: Yes, I'm not very clear with that, sorry. Maybe we should write a guideline on when to add/remove an author from a file (I think there's currently any, right?). The criteria I followed is: 
`sema.h` I did some significant additions but most of the code remained, so I added myself. In `sema.c` the code changed completely, so there was no code left from the previous authors, neither the authors.

I didn't want to remove the copyright because this is the legal thingy... but maybe I should remove it as well. Or do you prefer that I keep all authors and all copyrights?

File: sys/include/sema.h (hunk 9 above)
Comment (lebrush) [inline sys/include/sema.h]: Good point, actually it can only be initialized with a value different from 0. In order to initialize with 0 I need to initialize the mutex as locked, and this requires a change in the `mutex.h` file to make it clean (see #6210).

If that get's merged fast then I'll update this PR, if this PR gets merged first, I'll create a new one :-) The only addition would be:

```C
/**
 * @brief   Creates 0 initialized semaphore statically.
 * @return  Statically initialized semaphore.
 */
#define SEMA_CREATE_EMPTY()        { (value), SEMA_OK, MUTEX_INIT_LOCKED }
```

File: sys/include/sema.h [hunk 11]
```diff
@@ -33,23 +35,35 @@ extern "C" {
 /**
  * @brief   Creates semaphore statically.
  *
- * @param[in] value Initial value for the semaphore.
+ * @param[in] value Initial value for the semaphore (can't be 0). For a 0
+ *                  initialized semaphore @see SEMA_CREATE_EMPTY
```
Comment (miri64) [inline sys/include/sema.h]: 1. `SEMA_CREATE_EMPTY` does not exist
2. I don't understand the reasoning behind this restriction.

Comment (lebrush) [inline sys/include/sema.h]: See https://github.com/RIOT-OS/RIOT/pull/6155#discussion_r92351261 the restriction is that when the semaphore value is 0 the mutex needs to be locked.

Comment (miri64) [inline sys/include/sema.h]: Ah sorry, github hid that from me.

**Comment by miri64** [comment]
@lebrush wrote in a hidden comment

> If that get's merged fast then I'll update this PR, if this PR gets merged first, I'll create a new one :-) The only addition would be:
>
>     /**
>      * @brief   Creates 0 initialized semaphore statically.
>      * @return  Statically initialized semaphore.
>      */
>     #define SEMA_CREATE_EMPTY()        { (value), SEMA_OK, MUTEX_INIT_LOCKED }

I would call the the macro something like `SEMA_CREATE_LOCKED` since a semaphore isn't really something that is "filled" (i.e. it can't be empty).

File: sys/sema/sema.c (hunk 7 above)
Comment (miri64) [inline sys/sema/sema.c]: I don't really care... Just thought it looked quaint that the copyright was amended and the authors stripped ;-)

**Comment by lebrush** [comment]
#6210 has been merged. I added the locked initialization.

**Comment by miri64** [comment]
Since #6210 has been merged: can you rebase? (but not squash).

**Comment by lebrush** [comment]
Your wishes, my commands ;-) done! now there's an issue with the `list_remove` and some platforms... (see #6211) I'll have a look later, any hint? Maybe by just changing the return value of `list_remove` to `void`...

**Comment by miri64** [comment]
Aand #6158 got merged.

**Comment by lebrush** [comment]
Super cool! :dancer: I just rebased and squashed this one as well
@miri64 can you approve the changes and set the ready for ci, please? :-) thanks!

**Comment by lebrush** [comment]
ping @miri64

File: sys/include/sema.h [hunk 12]
```diff
@@ -58,92 +78,76 @@ typedef struct {
...
+ * Destroying a semaphore upon which other threads are currently blocked
+ * will wake the other threads causing the @ref sema_wait (or
+ * @ref sema_wait_timed) to return error (-ECANCELLED).
+ *
  * @param[in] sema  The semaphore to destroy.
  *
  * @return  0 on success.
- * @return  -EINVAL, if semaphore is invalid.
  */
 int sema_destroy(sema_t *sema);
```
Comment (miri64) [inline sys/include/sema.h]: Might be problematic, since it is an API change, but since we only have one state after return it might make sense to set the return type to `void`.

File: sys/include/sema.h [hunk 13]
```diff
@@ -58,92 +78,76 @@ typedef struct {
  * @param[in] value Initial value for the semaphore.
  *
  * @return  0 on success.
- * @return  -EINVAL, if semaphore is invalid.
```
Comment (miri64) [inline sys/include/sema.h]: Might be problematic, since it is an API change, but since we only have one state after return it might make sense to set the return type to void.

File: sys/include/sema.h [hunk 14]
```diff
@@ -58,92 +78,76 @@ typedef struct {
...
-/**
- * @brief   Wait for a semaphore being posted (without timeout, dropping spurious messages).
+ * @brief   Test if the semaphore is posted
  *
- * @param[in]  sema A semaphore.
+ * @pre `(sema != NULL)`
  *
- * @return  0 on success
- * @return  -EINVAL, if semaphore is invalid.
+ * @return  >= 0, on success number of left posts
```
Comment (miri64) [inline sys/include/sema.h]: Weird sentence structure

**Comment by miri64** [comment]
Sorry, lost track of this one :-/

File: sys/include/sema.h (hunk 13 above)
Comment (lebrush) [inline sys/include/sema.h]: About this you already complained earlier ([here](https://github.com/RIOT-OS/RIOT/pull/6155#discussion_r89315071) and [here](https://github.com/RIOT-OS/RIOT/pull/6155#discussion_r92342656)). The old return codes were restored (except if they made no sense anymore) and we aded a `@pre` tag. In any case this code is only used inside the posix semaphore and lwip and both have been adapted and tested respectively.

File: sys/include/sema.h (hunk 12 above)
Comment (lebrush) [inline sys/include/sema.h]: I left it to keep the old return values meaningful. This implementation can't fail here so the return value makes no difference, but the same applies to the creation of the semaphore.

I'm happy to change it to void... I've no strong opinion about this one... your call.

File: sys/include/sema.h (hunk 14 above)
Comment (lebrush) [inline sys/include/sema.h]: Me English no native. Will fix ;-)

**Comment by lebrush** [comment]
@miri64 I changed `sema_try_wait` to return 0 on success, the other situation makes no sense and this way all the API is consistent.

Regarding the return values for `sema_create` and `sema_destroy`. they are used in the POSIX implementation and in the lwip adaption. I'd leave them in even if they are always 0 just in case the implementation ever changes.

Let me know if this is Ok for you and I'll squash :-)

File: sys/include/sema.h (hunk 13 above)
Comment (miri64) [inline sys/include/sema.h]: > About this you already complained earlier (here [1] and here [2]). 

Mh, I would argue that [1] was about the change of the value itself and [2] was about the undocumented precondition (which is a different scope than return values because an unmet precondition leads to a crash, while an error reported by a return-value can be handled gracefully). However the comment in OP of this conversation was about arguing to remove, because according to this documentation the function only has one possible return value (`0`) and will always succeed, so having a return value seems redundant.

File: sys/include/sema.h (hunk 12 above)
Comment (miri64) [inline sys/include/sema.h]: `void` would minimize the code which in our use case is priority no. 1 ;-). So please go ahead.

**Comment by lebrush** [comment]
I made them in separate commits for you to easy spot them. Gonna squash :-)

**Comment by miri64** [comment]
The [error Jenkins reports](https://riot-ci.inet.haw-hamburg.de/job/RIOT/job/PR-6155/6/artifact/) was my bad in #5937 and is revealed, since you removed the include of `timex.h` from `sema.h`. Can you please squash the following patch onto the `pkg/lwip` commit:

```diff
diff --git a/pkg/lwip/contrib/sock/lwip_sock.c b/pkg/lwip/contrib/sock/lwip_sock.c
index 8ae799b..20b5faa 100644
--- a/pkg/lwip/contrib/sock/lwip_sock.c
+++ b/pkg/lwip/contrib/sock/lwip_sock.c
@@ -19,6 +19,7 @@
 #include "net/ipv4/addr.h"
 #include "net/ipv6/addr.h"
 #include "net/sock.h"
+#include "timex.h"
 
 #include "lwip/err.h"
 #include "lwip/ip.h"
```

**Comment by lebrush** [comment]
Thanks for being so responsive :-) just added your patch.
Btw is there a way to check the output of Jenkins quickly without downloading the artifacts? (the fancy new interface seems not to work)

**Comment by miri64** [comment]
> Btw is there a way to check the output of Jenkins quickly without downloading the artifacts? (the fancy new interface seems not to work)

That you have to ask @cgundogan or @smlng. My browser (chrome) just opens the .log files instead of downloading them

**Comment by lebrush** [comment]
mmm locally it works, just rebased to trigger Mr. Murdock again.

```
arduino_hello-world:nucleo-f030:
Building application "arduino_hello-world" for "nucleo-f030" with MCU "stm32f0".

/data/riotbuild/examples/arduino_hello-world/_sketches.cpp:81:5: error: redefinition of 'int ledPin'
 int ledPin = ARDUINO_LED;
     ^~~~~~
/data/riotbuild/examples/arduino_hello-world/_sketches.cpp:18:5: note: 'int ledPin' previously defined here
 int ledPin = ARDUINO_LED;
     ^~~~~~
(...)
```

**Comment by lebrush** [comment]
@miri64 green lights everywhere!! :tada:

**Comment by miri64** [comment]
Then go.
